scripts/main/lychee.js

```diff
@@ -639,6 +882,7 @@ lychee.setMode = function (mode) {
 	if (mode === "public") {
 		lychee.publicMode = true;
 	} else if (mode === "view") {
+		// TODO: It seems as if this method is never called with `mode === "view"`. This might be dead code. Can it be removed?
```

Comment:
That's because you recently removed the code that _did_ call it. Here it is in 4.4.0: https://github.com/LycheeOrg/Lychee-front/blob/a0713054c5b1d26ba80a470312d06805cde36343/scripts/main/album.js#L139-L144

The front end doesn't currently handle that case (of a public photo in a private album) correctly...

Reply:
Oh, damn it, my fault. So, I must really investigate what caused that `"Warning: Album private!"` message before the re-factoring and what is sent now instead. :-(

Reply:
Yeah, it's an interesting question really, how this _should_ be handled. We don't really _need_ the album to display a public photo, so why are we even attempting to load it? We do that so that previous/next/back work as expected. If the album is private though, should the server even send the album ID at all in response to `Photo::get`? I feel that it should set `album_id` to `null` in that case...

Reply:
> If the album is private though, should the server even send the album ID at all in response to Photo::get? I feel that it should set album_id to null in that case...

That would be one option, but this would probably also need some adoption on the front-end code so that the front-end does not accidentally believe that the photo is in the unsorted album. Moreover, it will again badly interfere with other types of client which cache the results in a local DB. I don't like fiddling with the "ground truth".

But anyway, now I know where I have to look and what the code was meant to do. Your insights are really helpful.

Reply:
> Moreover, it will again badly interfere with other types of client which cache the results in a local DB. I don't like fiddling with the "ground truth".

I think I agree with kamil4 here. Separating unsorted vs not provided is a consideration though. If that was accomplished it would remove any caching issues. My main view here is always "make Lychee work" as the primary goal though.

Reply:
> My main view here is always "make Lychee work" as the primary goal though.

Nobody disagree with that. However, it worked before without the trick of replacing the `album_id` by something differently. It worked in the past, because the hard-coded error message `""Warning: Album private!"` was caught and used as a sentinel condition to follow a different code path.

I haven't checked it, but I assume it should be a proper `403 FORBIDDEN` HTTP status now. So the only thing which I need to take care of during the 2nd half of the re-factoring is to correctly handle a `403` response and then follow the same code path as before. This would restore the old behavior but neither requires to replace `album_id` by some fake value nor relies on hard-coded error strings.

Reply:
@kamil4 wrote in [comment 797086084](https://github.com/LycheeOrg/Lychee-front/pull/282#discussion_r797086084):

> That's because you recently removed the code that _did_ call it. Here it is in 4.4.0:
> 
> https://github.com/LycheeOrg/Lychee-front/blob/a0713054c5b1d26ba80a470312d06805cde36343/scripts/main/album.js#L139-L144

I checkout version 4.3.6 and tried to find out when the backend replies with the string `"Warning: Album private!"`. The string is returned by

https://github.com/LycheeOrg/Lychee/blob/v4.3.6/app/Http/Middleware/ReadCheck.php#L42-L55

However, I was not able to trigger this code path for a single public photo in a private album. If a user opens a photo via `https://my-domain.tld/view?p=<photo-id>` the front-end simply does not try to get an album.

Of course, the code path is taken, if the user tries to open a private album, but that is a different story.

@kamil4  wrote in [comment 797086084](https://github.com/LycheeOrg/Lychee-front/pull/282#discussion_r797086084):

> The front end doesn't currently handle that case (of a public photo in a private album) correctly...

I believe I fixed this problem on my current, local development branch. There was some broken JS code in `scripts/view/main.js`, but that has been totally unrelated to `mode === "view"`. I am still convinced that the condition above is dead code. In particular, if a user browses `https://my-domain.tld/view?p=<photo-id>`, the JS does not even include the script file `scripts/main/lychee.js` but is built from `script/view/main.js`.

@kamil4 You said that the front-end doesn't currently handle the case of a public photo in a private album correctly. Could you give me a precise step-by-step introduction which use case is not handled correctly so that I can test it on my development branch? In particular: Is the user unauthenticated or authenticated? Do you somehow "browse" to the photo or do you invoke it by direct URL? Which URL are you using `/r/view?p=<photo-id>`, or `/r/<photo-id>`, or `gallery#<photo-id>`?

Reply:
Sorry, I somehow missed this response from you... I believe the case that wasn't handled correctly was `#<album_id>/<photo_id>` in public mode, i.e., a simple copy-paste of the URL from the address bar of the owner, for a public photo in a private album. But I just checked and it now works as I expect.